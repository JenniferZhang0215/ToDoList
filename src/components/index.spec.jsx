import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import App from "./index";
import List from "./list";
import Status from "./status";

configure({ adapter: new Adapter() });
describe("<App />", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe("when inputArray is empty", () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ inputArray: [] });

    it("should render default snapshot", () => {
      expect(wrapper).toMatchSnapshot("default");
    });

    it("should render correctly", () => {
      expect(wrapper).toMatchSnapshot("default");
    });

    it("should render No ToDos Here", () => {
      const title = wrapper.find("h3");
      expect(title.text()).toEqual("No ToDos Here");
    });

    describe("input", () => {
      const input = wrapper.find("input");
      const e = { target: { value: "fake input" }, keyCode: 13 };
      it("should call handleChange when input is changed", () => {
        const id = wrapper.state("id");
        input.simulate("change", e);
        expect(wrapper.state("newData")).toEqual({
          key: id,
          text: e.target.value,
          edit: false,
          status: false,
          priority: 0,
        });
      });

      it("should call handleKeyDown", () => {
        const instance = wrapper.instance();
        jest.spyOn(instance, "handleClick");
        input.simulate("keyDown", e);
        expect(instance.handleClick).toHaveBeenCalled();
      });

      it("should render correct placeholder", () => {
        expect(input.props().placeholder).toEqual("Add New To Do Item");
      });

      it("should pass correct value", () => {
        expect(input.props().value).toEqual(wrapper.state("newData").text);
      });
    });

    describe("button", () => {
      const button = wrapper.find("button");
      it("should call handleClick if add button is clicked", () => {
        const mockNewData = {
          key: 0,
          text: "fake text",
          edit: false,
          status: false,
          priority: 0,
        };
        wrapper.setState({
          newData: mockNewData,
          inputArray: [],
          id: 1,
        });
        button.simulate("click");
        expect(wrapper.state("newData")).toEqual({
          edit: false,
          key: 1,
          priority: 0,
          status: false,
          text: "",
        });
        expect(wrapper.state("inputArray")).toEqual([mockNewData]);
        expect(wrapper.state("id")).toEqual(2);
      });

      it("should render text Add", () => {
        expect(button.text()).toEqual("Add");
      });
    });
  });

  describe("when inputArray is not empty", () => {
    const wrapper = shallow(<App />);
    const initialInputArray = [
      { edit: false, key: 0, priority: 0, status: false, text: "fake text 1" },
      { edit: false, key: 1, priority: 0, status: false, text: "fake text 2" },
    ];
    wrapper.setState({
      inputArray: initialInputArray,
    });

    it("should render correctly", () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe("filteredInputArray", () => {
      const instance = wrapper.instance();

      it("when selectedStatus is All", () => {
        expect(instance.filteredInputArray).toEqual(
          wrapper.state("inputArray")
        );
      });

      it("when selectedStatus is Done", () => {
        wrapper.setState({ selectedStatus: "Done" });
        expect(instance.filteredInputArray).toEqual([]);
      });

      it("when selectedStatus is Processing", () => {
        wrapper.setState({ selectedStatus: "Processing" });
        expect(instance.filteredInputArray).toEqual(initialInputArray);
      });

      it("when searchText exists", () => {
        wrapper.setState({ searchText: "fake text 1" });
        expect(instance.filteredInputArray).toEqual([
          {
            edit: false,
            key: 0,
            priority: 0,
            status: false,
            text: "fake text 1",
          },
        ]);
      });
    });

    describe("search bar input", () => {
      const input = wrapper.findWhere((node) => node.key() === "search-bar");

      it("should call handleSearch when input is changed", () => {
        const e = { target: { value: "fake search content" } };
        input.simulate("change", e);
        expect(wrapper.state("searchText")).toEqual(e.target.value);
      });

      it("should have correct placeholder value", () => {
        expect(input.props().placeholder).toEqual("Search The To Do List");
      });
    });

    describe("<List />", () => {
      const listComponent = wrapper.find(List);

      it("should have correct inputArray value", () => {
        expect(listComponent.props().inputArray).toEqual(initialInputArray);
      });

      it("should assign handleDelete function as onDelete", () => {
        const item = {
          edit: false,
          key: 1,
          priority: 0,
          status: false,
          text: "fake text 2",
        };
        listComponent.simulate("delete", item);
        expect(wrapper.state("inputArray")).toEqual([
          {
            edit: false,
            key: 0,
            priority: 0,
            status: false,
            text: "fake text 1",
          },
        ]);
      });

      it("should assign handleEdit function as onEdit", () => {
        const item = {
          edit: false,
          key: 0,
          priority: 0,
          status: false,
          text: "fake text 1",
        };
        listComponent.simulate("edit", item);
        expect(wrapper.state("inputArray")).toEqual([
          {
            edit: true,
            key: 0,
            priority: 0,
            status: false,
            text: "fake text 1",
          },
        ]);
      });

      it("should assign handleSave function as onEdit", () => {
        const item = {
          edit: true,
          key: 0,
          priority: 0,
          status: false,
          text: "fake text 1",
        };
        const editedText = "edited fake text 1";
        listComponent.simulate("save", item, editedText);
        expect(wrapper.state("inputArray")).toEqual([
          {
            edit: false,
            key: 0,
            priority: 0,
            status: false,
            text: editedText,
          },
        ]);
      });

      it("should assign handleCancel function as onEdit", () => {
        const item = {
          edit: true,
          key: 0,
          priority: 0,
          status: false,
          text: "fake text",
        };
        wrapper.setState({ inputArray: [item] });
        listComponent.simulate("cancel", item);
        expect(wrapper.state("inputArray")).toEqual([
          {
            edit: false,
            key: 0,
            priority: 0,
            status: false,
            text: "fake text",
          },
        ]);
      });

      it("should assign handleFinish function as onEdit", () => {
        const item = {
          edit: false,
          key: 0,
          priority: 0,
          status: false,
          text: "fake text",
        };
        wrapper.setState({ inputArray: [item] });
        listComponent.simulate("finish", item);
        expect(wrapper.state("inputArray")).toEqual([
          {
            edit: false,
            key: 0,
            priority: 0,
            status: true,
            text: "fake text",
          },
        ]);
      });

      it("should assign handleEditPriority function as onEdit", () => {
        const item = {
          edit: false,
          key: 0,
          priority: 0,
          status: false,
          text: "fake text",
        };
        const e = { target: { value: 1 } };
        wrapper.setState({ inputArray: [item] });
        listComponent.simulate("editPriority", e, item);
        expect(wrapper.state("inputArray")).toEqual([
          {
            edit: false,
            key: 0,
            priority: e.target.value,
            status: false,
            text: "fake text",
          },
        ]);
      });
    });

    describe("<Status />", () => {
      const statusComponent = wrapper.find(Status);

      it("should have correct statusArray value", () => {
        expect(statusComponent.props().statusArray).toEqual(
          wrapper.state("statusArray")
        );
      });

      it("should call statusArray", () => {
        const status = "Done";
        statusComponent.simulate("status", status);
        expect(wrapper.state("selectedStatus")).toEqual(status);
      });
    });
  });
});
