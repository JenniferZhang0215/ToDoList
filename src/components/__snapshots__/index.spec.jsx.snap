// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> when inputArray is empty should render correctly: default 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          No ToDos Here
        </h3>,
        <input
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Add New To Do Item"
          value=""
        />,
        <button
          className="btn-success btn-sm m-2"
          disabled=""
          onClick={[Function]}
        >
          Add
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No ToDos Here",
        },
        "ref": null,
        "rendered": "No ToDos Here",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "onKeyDown": [Function],
          "placeholder": "Add New To Do Item",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add",
          "className": "btn-success btn-sm m-2",
          "disabled": "",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            No ToDos Here
          </h3>,
          <input
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Add New To Do Item"
            value=""
          />,
          <button
            className="btn-success btn-sm m-2"
            disabled=""
            onClick={[Function]}
          >
            Add
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "No ToDos Here",
          },
          "ref": null,
          "rendered": "No ToDos Here",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "Add New To Do Item",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add",
            "className": "btn-success btn-sm m-2",
            "disabled": "",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<App /> when inputArray is empty should render default snapshot: default 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          No ToDos Here
        </h3>,
        <input
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Add New To Do Item"
          value=""
        />,
        <button
          className="btn-success btn-sm m-2"
          disabled=""
          onClick={[Function]}
        >
          Add
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No ToDos Here",
        },
        "ref": null,
        "rendered": "No ToDos Here",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "onKeyDown": [Function],
          "placeholder": "Add New To Do Item",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add",
          "className": "btn-success btn-sm m-2",
          "disabled": "",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            No ToDos Here
          </h3>,
          <input
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Add New To Do Item"
            value=""
          />,
          <button
            className="btn-success btn-sm m-2"
            disabled=""
            onClick={[Function]}
          >
            Add
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "No ToDos Here",
          },
          "ref": null,
          "rendered": "No ToDos Here",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "Add New To Do Item",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add",
            "className": "btn-success btn-sm m-2",
            "disabled": "",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<App /> when inputArray is not empty should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          onChange={[Function]}
          placeholder="Search The To Do List"
          value=""
        />,
        <List
          inputArray={
            Array [
              Object {
                "edit": false,
                "key": 0,
                "priority": 0,
                "status": false,
                "text": "fake text 1",
              },
              Object {
                "edit": false,
                "key": 1,
                "priority": 0,
                "status": false,
                "text": "fake text 2",
              },
            ]
          }
          onCancel={[Function]}
          onDelete={[Function]}
          onEdit={[Function]}
          onEditPriority={[Function]}
          onFinish={[Function]}
          onSave={[Function]}
        />,
        <Status
          onStatus={[Function]}
          selectedStatus="All"
          statusArray={
            Array [
              "All",
              "Processing",
              "Done",
            ]
          }
        />,
        <div>
          <input
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Add New To Do Item"
            value=""
          />
          <button
            className="btn-success btn-sm m-2"
            onClick={[Function]}
          >
            Add
          </button>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "search-bar",
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Search The To Do List",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "inputArray": Array [
            Object {
              "edit": false,
              "key": 0,
              "priority": 0,
              "status": false,
              "text": "fake text 1",
            },
            Object {
              "edit": false,
              "key": 1,
              "priority": 0,
              "status": false,
              "text": "fake text 2",
            },
          ],
          "onCancel": [Function],
          "onDelete": [Function],
          "onEdit": [Function],
          "onEditPriority": [Function],
          "onFinish": [Function],
          "onSave": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onStatus": [Function],
          "selectedStatus": "All",
          "statusArray": Array [
            "All",
            "Processing",
            "Done",
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Add New To Do Item"
              value=""
            />,
            <button
              className="btn-success btn-sm m-2"
              onClick={[Function]}
            >
              Add
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "add-input",
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "onKeyDown": [Function],
              "placeholder": "Add New To Do Item",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add",
              "className": "btn-success btn-sm m-2",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            onChange={[Function]}
            placeholder="Search The To Do List"
            value=""
          />,
          <List
            inputArray={
              Array [
                Object {
                  "edit": false,
                  "key": 0,
                  "priority": 0,
                  "status": false,
                  "text": "fake text 1",
                },
                Object {
                  "edit": false,
                  "key": 1,
                  "priority": 0,
                  "status": false,
                  "text": "fake text 2",
                },
              ]
            }
            onCancel={[Function]}
            onDelete={[Function]}
            onEdit={[Function]}
            onEditPriority={[Function]}
            onFinish={[Function]}
            onSave={[Function]}
          />,
          <Status
            onStatus={[Function]}
            selectedStatus="All"
            statusArray={
              Array [
                "All",
                "Processing",
                "Done",
              ]
            }
          />,
          <div>
            <input
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Add New To Do Item"
              value=""
            />
            <button
              className="btn-success btn-sm m-2"
              onClick={[Function]}
            >
              Add
            </button>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "search-bar",
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Search The To Do List",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "inputArray": Array [
              Object {
                "edit": false,
                "key": 0,
                "priority": 0,
                "status": false,
                "text": "fake text 1",
              },
              Object {
                "edit": false,
                "key": 1,
                "priority": 0,
                "status": false,
                "text": "fake text 2",
              },
            ],
            "onCancel": [Function],
            "onDelete": [Function],
            "onEdit": [Function],
            "onEditPriority": [Function],
            "onFinish": [Function],
            "onSave": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onStatus": [Function],
            "selectedStatus": "All",
            "statusArray": Array [
              "All",
              "Processing",
              "Done",
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onChange={[Function]}
                onKeyDown={[Function]}
                placeholder="Add New To Do Item"
                value=""
              />,
              <button
                className="btn-success btn-sm m-2"
                onClick={[Function]}
              >
                Add
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "add-input",
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "onKeyDown": [Function],
                "placeholder": "Add New To Do Item",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add",
                "className": "btn-success btn-sm m-2",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
